skglm/solvers/gram_cd.py

```diff
@@ -53,85 +43,88 @@ def gram_cd_solver(X, y, penalty, max_iter=100, w_init=None,
     """
-    n_samples, n_features = X.shape
-
-    if issparse(X):
-        scaled_gram = X.T.dot(X)
-        scaled_gram = scaled_gram.toarray() / n_samples
-        scaled_Xty = X.T.dot(y) / n_samples
-    else:
-        scaled_gram = X.T @ X / n_samples
-        scaled_Xty = X.T @ y / n_samples
-    # TODO potential improvement: allow to pass scaled_gram (e.g. for path computation)
-
-    scaled_y_norm2 = np.linalg.norm(y)**2 / (2*n_samples)
 
-    all_features = np.arange(n_features)
-    stop_crit = np.inf  # prevent ref before assign
-    p_objs_out = []
-
-    w = np.zeros(n_features) if w_init is None else w_init
-    grad = - scaled_Xty if w_init is None else scaled_gram @ w_init - scaled_Xty
-    opt = penalty.subdiff_distance(w, grad, all_features)
-
-    if use_acc:
-        if greedy_cd:
-            warnings.warn(
-                "Anderson acceleration does not work with greedy_cd, set use_acc=False",
-                UserWarning)
-        accelerator = AndersonAcceleration(K=5)
-        w_acc = np.zeros(n_features)
-        grad_acc = np.zeros(n_features)
-
-    for t in range(max_iter):
-        # check convergences
-        stop_crit = np.max(opt)
-        if verbose:
-            p_obj = (0.5 * w @ (scaled_gram @ w) - scaled_Xty @ w +
-                     scaled_y_norm2 + penalty.value(w))
-            print(
-                f"Iteration {t+1}: {p_obj:.10f}, "
-                f"stopping crit: {stop_crit:.2e}"
-            )
-
-        if stop_crit <= tol:
-            if verbose:
-                print(f"Stopping criterion max violation: {stop_crit:.2e}")
-            break
-
-        # inplace update of w, grad
-        opt = _gram_cd_epoch(scaled_gram, w, grad, penalty, greedy_cd)
-
-        # perform Anderson extrapolation
-        if use_acc:
-            w_acc, grad_acc, is_extrapolated = accelerator.extrapolate(w, grad)
-
-            if is_extrapolated:
-                # omit constant term for comparison
-                p_obj_acc = (0.5 * w_acc @ (scaled_gram @ w_acc) - scaled_Xty @ w_acc +
-                             penalty.value(w_acc))
-                p_obj = 0.5 * w @ (scaled_gram @ w) - scaled_Xty @ w + penalty.value(w)
-                if p_obj_acc < p_obj:
-                    w[:] = w_acc
-                    grad[:] = grad_acc
-
-        # store p_obj
-        p_obj = (0.5 * w @ (scaled_gram @ w) - scaled_Xty @ w + scaled_y_norm2 +
-                 penalty.value(w))
-        p_objs_out.append(p_obj)
-    return w, np.array(p_objs_out), stop_crit
+    def __init__(self, max_iter=100, use_acc=True, greedy_cd=True, tol=1e-4,
+                 fit_intercept=True, warm_start=False, verbose=0):
+        self.max_iter = max_iter
+        self.use_acc = use_acc
+        self.greedy_cd = greedy_cd
+        self.tol = tol
+        self.fit_intercept = fit_intercept
+        self.warm_start = warm_start
+        self.verbose = verbose
+
+    def solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
```

Comment:
We don't pass in ``Xw_init``.
IMO, we should hint at this.

Reply:
I stick to the definition in the abstract class, but indeed I don't use it in the function body. WDYT?

Reply:
I added an small comment as hint